src/psyclone/gocean1p0.py

```diff
@@ -3278,12 +3291,19 @@ def _read_from_device_routine(self, f2pygen_module=None, psyir=None):
         # Rename subroutine
         subroutine.name = subroutine_name
 
-        # If a f2pygen modu is provided insert a PSyIRGen node
+        # If a f2pygen module is provided insert a PSyIRGen node
         if f2pygen_module:
             f2pygen_module.add(PSyIRGen(f2pygen_module, subroutine))
 
-        # If a psyir is provided insert the routine as a Container child
+        # If PSyIR is provided insert the routine as a child of the parent
+        # Container
         if psyir:
+            if not psyir.ancestor(Container):
+                raise GenerationError(
+                    "The GOACCEnterDataDirective can only be generated/lowered"
```

Comment:
Cut-n-paste error? We're attempting to add an implementation of the `read_openacc` subroutine, not a GOACCEnterDataDirective.

Reply:
I think the error message is fine, we add a read_openacc subroutine if it's not already there when we find a GOACCEnterDataDirective and for this we need a top level container (the public API methods that call this routine are the GOACCEnterDataDirective  gen_code and lower_to_language_level)

Reply:
Ah I see - thanks.